shards/aggregate.go

```diff
@@ -0,0 +1,137 @@
+package shards
+
+import (
+	"sync"
+	"time"
+
+	"github.com/sourcegraph/zoekt"
+	"github.com/sourcegraph/zoekt/stream"
+)
+
+// collectSender is a sender that will aggregate results. Once sending is
+// done, you call Done to returned the aggregated result which are ranked.
+type collectSender struct {
+	aggregate          *zoekt.SearchResult
+	maxDocDisplayCount int
+}
+
+func newCollectSender(opts *zoekt.SearchOptions) *collectSender {
+	return &collectSender{
+		aggregate: &zoekt.SearchResult{
+			RepoURLs:      map[string]string{},
+			LineFragments: map[string]string{},
+		},
+		maxDocDisplayCount: opts.MaxDocDisplayCount,
+	}
+}
+
+func (c *collectSender) Send(r *zoekt.SearchResult) {
+	c.aggregate.Stats.Add(r.Stats)
+
+	if len(r.Files) > 0 {
+		c.aggregate.Files = append(c.aggregate.Files, r.Files...)
+
+		for k, v := range r.RepoURLs {
+			c.aggregate.RepoURLs[k] = v
+		}
+		for k, v := range r.LineFragments {
+			c.aggregate.LineFragments[k] = v
+		}
+	}
+
+	// The priority of our aggregate is the largest priority we collect.
+	if c.aggregate.Priority < r.Priority {
+		c.aggregate.Priority = r.Priority
+	}
+
+	// We receive monotonically decreasing values, so we update on every call.
+	//
+	// TODO we need to think about only updating this if smaller. But then we
+	// need to handle no events not sending invalid Progress.
+	c.aggregate.MaxPendingPriority = r.MaxPendingPriority
+}
+
+// Done returned the aggregated result. Before returning them the files are
+// ranked and truncated according to the input SearchOptions.
+//
+// Note: copyFiles is called on the returned value. Additionally Done can only
+// be called once.
+func (c *collectSender) Done() *zoekt.SearchResult {
+	// TODO what happens if we never receive a result? We need to ensure we
```

Comment:
streamSearch always returns a Progress event per shard, even if don't find anything.

Reply:
thinking about the case were there are no shards. We can special case that in the layer above?

Reply:
as soon as we call getShards and get 0 shards we could return early. In Sourcegraph, once a Zoekt returns, we already remove the entry in the maxPendingMap for that Zoekt so that we don't keep waiting. IE we don't necessarily have to send progress if we don't have any shards.